python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1121,344 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
         Interpolation matrix.
     """
     return _interpolation_matrix(space0._cpp_object, space1._cpp_object)
+
+
+class SNESProblem:
+    def __init__(
+        self,
+        F: typing.Union[dolfinx.fem.Form, ufl.form.Form],
+        u: dolfinx.fem.Function,
+        J: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        P: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        bcs: typing.Optional[list[dolfinx.fem.DirichletBC]] = None,
+        form_compiler_options: typing.Optional[dict] = None,
+        jit_options: typing.Optional[dict] = None,
+    ):
+        """Class for constructing the residual and Jacobian constructors for a SNES problem.
+
+        Args:
+            F: The PDE residual F(u, v)
+            u: The unknown
+            bcs: List of Dirichlet boundary conditions
+            J: UFL representation of the Jacobian (Optional)
+            P: UFL representation of a preconditioner (Optional)
+            form_compiler_options: Options used in FFCx
+                compilation of this form. Run ``ffcx --help`` at the
+                command line to see all available options.
+            jit_options: Options used in CFFI JIT compilation of C
+                code generated by FFCx. See ``python/dolfinx/jit.py``
+                for all available options. Takes priority over all other
+                option values.
+        """
+        self._comm = u.function_space.mesh.comm
+        form_compiler_options = {} if form_compiler_options is None else form_compiler_options
+        jit_options = {} if jit_options is None else jit_options
+
+        self._L = _create_form(
+            F, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+        if J is None:
+            V = u.function_space
+            du = ufl.TrialFunction(V)
+            J = ufl.derivative(F, u, du)
+        self._a_prec = (
+            None
+            if P is None
+            else _create_form(
+                P, form_compiler_options=form_compiler_options, jit_options=jit_options
+            )
+        )
+        self._a = _create_form(
+            J, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+        self.bcs = bcs
+        self.u = u
+
+    @property
+    def comm(self):
```

Comment:
Is this property needed? If yes, should it come from the SNES object?

Reply:
It is needed to construct the SNES object.
We don't easily store the communicator within compiled forms, especially when it can be a form, a list of forms or a nested list of forms.